content/articles/2017-12-14-styler-1-0-0.Rmd

```diff
@@ -0,0 +1,234 @@
+---
+title: styler 1.0.0
+author: Lorenz Walthert, Kirill Müller
+date: '2017-12-14'
+slug: styler-1.0.0
+categories:
+  - package
+photo:
+  url: https://unsplash.com/photos/mpdIPhYqZ4Y
+  author: Heng Films
+description: >
+  styler, a package that can format your code according to the tidyverse style guide, is now on CRAN.
+
+---
+
+```{r setup, include = FALSE}
+knitr::opts_chunk$set(echo = TRUE, comment = "#>")
+knitr::knit_engines$set(list(
+  styler = function(options) {
+    options$comment <- ""
+    knitr::engine_output(
+      options,
+      
+      c("# Before", options$code),
+      c("# After", styler::style_text(options$code))
+    )
+  }
+))
+
+```
+
+
+We're pleased to announce the release of [styler 1.0.0](https://github.com/r-lib/styler/releases/tag/v1.0.0). styler is a source code formatter - a package to format R code according to a style guide. We've implemented the [tidyverse style guide](http://style.tidyverse.org), so you can format your code according to it. A coherent style is important for consistency and legibility. Just as it is important to putspcesbetweenwords. 
+You can install styler from CRAN:
+
+```{r, eval = FALSE}
+install.packages("styler")
+```
+
+styler can style text, single files, packages and entire
+R source trees with the following functions:
+
+* `style_text()` styles a character vector.
+* `style_file()` styles R and Rmd files.
+* `style_dir()` styles all R and/or Rmd files in a directory.
+* `style_pkg()` styles the source files of an R package.
+* An RStudio Addin that styles the active file R or Rmd file, the current 
+  package or the highlighted code.
+
+## styling options
+
+A distinguishing feature of styler is its flexibility. The following options are available:
+
+* `scope`. What is to be styled, i.g. just spacing or also indention, line breaks, tokens?
+* `strict`. How strict should be rules be applied?
+* `math_token_spacing`. Granular control for spacing around math tokens.
+* `reindention`. Allows to set the indention of certain regex patterns manually.
+
+We will briefly describe all of them below.
```

Comment:
It seems like restyling is a natural pairing with version control. That is, if you're doing to do this and you use version control ... definitely commit right before you restyle! Now that I think of it, it's not clear to me if styler actually changes files or highlighted code or just returns restyled code? Maybe a few sentences about the workflow(s) would be helpful.

Reply:
Good point, maybe we should raise that. I am not sure whether I understand you correctly though. I think there are different approaches. One would be:

* Implement new feature by editing a code file.
* commit new feature
* apply styler
* inspect the results (and make sure the changes are as desired) and `git commit ... --amend` to make the committed feature "clean".

But that's a bit tedious. I either style right before I commit (and assume styling does not break anything, and if it should, I can always amend), or I only style occasionally (once every 20-40 commits, but then I style all files, e.g. with `style_pkg()`). I think if every second commit message is "applying styler", then the history becomes "cluttered" so I am not sure whether that is a good idea. What do you think?

Reply:
I just meant this: It's a pretty big deal that styler will edit files in place. That's worth pointing out explicitly. And it has obvious implications and synergy with version control.

Reply:
Ok, sure, we can do it like that.